### DISCUSSION TITLE
AppCheck with Play Integrity?

### DISCUSSION THREAD

**Discussion by pierroo** [invertase/react-native-firebase#6345] [Q&A] [answered]
Hi,

In the docs, there is only mention of safetyNet for AppCheck, but nothing about Play Integrity which seems to be the new way of Google to verify the authenticity of an app and they plan on deprecating SafetyNet.

Is there any plan to include Play Integrity in the future?

Bonus question: will Play Integrity work on huawei devices or anything that does not have google play services? :/

**Comment by mikehardy** [comment]
I can't see how that would ever work without Google Play Services (it seems to be deeply linked even down to the legalese? https://developer.android.com/google/play/integrity/terms#data-safety)

Anyway, I can see they added that about 2 months ago - https://firebase.google.com/support/release-notes/android#appcheck_v16-0-0, I haven't had a chance to wrap it up and expose it here and likely won't for a few more months. In the meantime if it's needed for your use cases I'll happily help merge a PR

**Reply by pierroo** [reply]
Thank you Mike for your answer as always.
Definitely NOT needed so far, although I just got worried when I saw all their warnings about SafeNet being deprecated soon and asking to use Play Integrity right away.

Which leads me to some more question if you'd allow me:

- Can SafetyNet work without google play services on the other hand? (my main concern being Huwaei device unable to use my app if they don't pass the AppCheck on android :/)
- so right now for android when we want to activate appcheck, this is the line I found online :
appCheck().activate('ignored', true)

Can you confirm this is correct? ('ignored', mainly; since I do want the true as second one)
Somehow it would be nice to add it to the documentation in the "activate" part, this information was quite buried in some stackoverflow post

And finally, do we actually need to add "implementation 'com.google.firebase:firebase-appcheck-safetynet:16.0.0'" in our app/build.gradle?
It's in the official documentation, but not clear either in the RNFirebase one if we need it?

For some reason I never had issue with any other RNFirebase integration, but seems like AppCheck is quite unclear yet.
(most probably because it just left Beta stage)

**Reply by mikehardy** [reply]
Here are the "GMS" (Play Services) requirements https://firebase.google.com/docs/android/android-play-services

AppCheck custom token provider can work on non-GMS devices, so perhaps someone has written an AppCheck provider that works with HMS somehow? Nothing popped up immediately though https://search.brave.com/search?q=appcheck+huawei+firebase 

Perhaps an AppCheck custom provider around SysIntegrity (https://developer.huawei.com/consumer/en/hms/huawei-safetydetectkit/) and a PR here that loaded it somehow? firebase-android-sdk may be open to a Huawei provider even, I'm not sure. Otherwise you would need a custom provider of some sort that issued something for Huawei devices or you'd never be able to turn on "enforcing" mode

> Can you confirm this is correct? ('ignored', mainly; since I do want the true as second one)

Yes

> Somehow it would be nice to add it to the documentation in the "activate" part, this information was quite buried in some stackoverflow post

Please hit the "edit" button at top right of every doc page if you see something confusing, our users have fresh eyes on confusing things, I certainly don't any more - even the most bizarre parts of the system appear normal to me now unfortunately

> And finally, do we actually need to add "implementation 'com.google.firebase:firebase-appcheck-safetynet:16.0.0'" in our app/build.gradle?

I would do exactly what is done in our repo `tests/` app with regard to dependencies. I don't think it's necessary but that's our reference app and the app-check stuff is pretty subtle *especially* the parts about debug tokens for CI tests and such

AppCheck is difficult. Not quite as difficult as getting for instance FCM working but it is difficult and I think partially yes because the number of users isn't high yet so the docs haven't been smoothed out via questions+PRs. Hopefully it will get there though, and hopefully this helps?

**Reply by pierroo** [reply]
Thank you for the links Mike.

But the first one is actually what worries me, it initially says that com.google.firebase:firebase-appcheck:16.0.0 doesn't need play services;
but then below you see that both safetyNet and play integrity require google play services.
And as far as I know, appcheck relies on either one of those to work.

Yes it also allows for a custom check provider, but anything that falls out of rnfirebase support scares me in case of issue.
But even then, if there is such a custom check provider, how would I use it? since the firebase console in appcheck only allows for safetynet and play integrity, how can I add the custom provider sha256; or perhaps I don't need to do so?

So, from the moment I activate appcheck in my RN code, I would need to detect if the device is a huawei one, and change the line accordingly?
For example:

// isHuawei from device detection
if (isHuawei) {
await appCheck().activate(customProvider, true)
} else {
await appCheck().activate('ignored', true)
}

but even then, the documentation clearly says "@param siteKeyOrProvider - This is ignored" so they wouldn't even care if I passed a customProvider there?

I will look into the tests/ app as well, thank you for this suggestion too :)

Edit:
I just checked your test repo (https://github.com/invertase/react-native-firebase/blob/main/tests/app.js , is that correct?), unfortunately although app-check is imported it is not used anywhere, nor is it mentioned in the app/build.gradle.

**Reply by mikehardy** [reply]
There is "AppCheck" the system. And there are "token providers" that provide the tokens. Debug Providers, specific providers for other things, etc

The current state of AppCheck is that *simply* linking it in does most of the work, in fact on iOS for the default provider there it does *all* the work, you just need to include the package and for production purposes you are done.

On android you need to call activate and that *only* checks build type, installing DebugProvider for debug builds, SafetyNet provider for release builds.

The code is trivial, you could/should check it all: 

![image](https://user-images.githubusercontent.com/782704/176229204-47964b40-77af-496d-b4e3-e35bcaa8f3fd.png)


https://github.com/invertase/react-native-firebase/blob/f9a50496bddab6a4bb740cd18dd7407d219fbda2/packages/app-check/android/src/main/java/io/invertase/firebase/appcheck/ReactNativeFirebaseAppCheckModule.java#L41-L85

When AppCheck was first released there was the debug provider and the SafetyNet provider and that's it, so this structure made sense.

Further, there is *no* way to configure new providers at run time that I know of. You install a provider at startup and that's it, so this structure was almost the only way to do it ?

Now that there are multiple providers for iOS and Android, a natural future extension is to allow users to specify via javascript APIs or in firebase.json the provider(s) to use (somehow?) and then implement a react-native-firebase specific "Provider Facade" that we can install in the underlying SDKs as the token provider, and switch in different concrete providers at run time.

Once that is done we can do what the SDKs expect at native startup, then after JS starts we can have APIs that detect Huawei for instance and swap in AppCheckHuaweiProvider (which does not exist) and call deleteToken/getToken to force the new state

Or same to switch from SafetyNet to PlayIntegity on Android or on iOS to switch to AppAttest or similar.

This is not the largest job but it's not a small job. Given the way AppCheck initialization works I can't think of another way to allow javascript-level platform checks / capability checks to allow people to use other providers

**Reply by pierroo** [reply]
Thank you for this very detailed answer.

I had such high hopes that appcheck was gonna be the life saver I expected against fake API requests from hackers outside of my mobile apps; unfortunately it seems that it would come at a cost... my huwaei userbase, which is most probably 10% of them all. :/

Maybe I will have to handle it completely differently; if Huawei then use recaptcha v3 and hit a different endpoint, otherwise use appcheck and hit the normal endpoint. that seems quite ugly, but could be a good compromise until firebase looks into it I guess? (if they ever plan to? Not sure they care about huawei users; or is it actually the other way around? ^_^)

**Reply by mikehardy** [reply]
I'm not sure they plan to, this is something that others can do if really needed as well
Your idea of splitting end points is interesting that could definitely work

For attacks you can also have a "penalty box" mode, where you exponentially backoff requests from same API etc to throttle bad actors but in a way where they continue trying, just less effectively maybe. That's a classic defense against floods - don't just deny because they'll try again, make it look like they're getting through but just really slowly so you tie up their resources instead of yours

**Reply by pierroo** [reply]
This API throttling, API IP limiter etc, are all already in place.

However, when it comes to expert hackers using OpenBullets/SilverBullets that generates new IP through proxies for each request with thousands of those per seconds, there is just not much you can do besides praying or closing your API entirely until they stop. :')

That is why Firebase AppCheck seemed so promising, because at least I could make sure that those requests came from the real mobile app and not from some shady third party tool. (this, unless they can actually misuse the appcheck token for as long as its TTL doesn't expire, but that we discussed on another topic :))

**Reply by mikehardy** [reply]
If you've already gone to those lengths, it may not be *that* much more difficult to do something quick-and-dirty in node_modules / persisted via patch-package where you

- write the AppCheckSysIntegrityProvider (no idea on the SysIntegrity APIs, hopefully easy to use? AppCheck android providers are open source for inspiration)
- modify our android app check code to check for non-debug builds if HMS classes are available (try to class.forName them ?) and if they are install that provider not the AppCheck provider 
- done?

Most (all?) of the code is structural - just binding one API to another and handling errors, it may not be too big a job

**Reply by Saran-Avinash** [reply]
Please anyone share if there is  any information written for how to make app checks with play integrity api possible in react-native

**Comment by RyadhB2** [comment]
Hello, I wanted to ask if there's any new update about the subject, ie. is it planned to replace SafetyNet or add Play integrity tin the future ?

**Reply by mikehardy** [reply]
There is no news. The clock is ticking for us (or a motivated contributor from the community) to add a Play Integrity provider for react-native-firebase though. We have a few more months at least

**Reply by RyadhB2** [reply]
Thanks a lot for the fast reply 🙏🙏. We have a few more months indeed and hopefully it will be added

**Reply by RyadhB2** [reply]
Hi, any updates please ?

**Reply by mikehardy** [reply]
@RyadhB2 please follow the associated pull request, it's nearly ready #6847 - that said, this should not affect you as you have been asking since December, and the only thing that happened today is *new* signups are disallowed. If you've been using AppCheck since December you are still more than able to use SafetyNet.

**Comment by mo-nizar** [comment]
is there any update on this issue, i am still not able to fetch app check token from react native android application

**Comment by mikehardy** [comment, marked as answer]
Android Play Integrity support has been released with react-native-firebase v17!

**Reply by ledinhthi** [reply]
HI @mikehardy mike
<img width="1136" alt="Screen Shot 2023-02-27 at 17 22 36" src="https://user-images.githubusercontent.com/25853138/221538181-40a52f84-8fb6-4ee3-b1fb-f44ca6553856.png">

I fetch token Android Play Integrity. but it show attestation 403 . So what should i do to get toklen?.
Thank you so much

**Comment by sachin3072004** [comment]
Hi Mike, With this change phone auth should work with the 'Play Integrity API' ? right ?

**Reply by mikehardy** [reply]
They are related but separate. It appears that phone auth may still require the device check API in order to avoid recaptcha - I apologies I don't have a moment to look right now but I *know* stackoverflow has a question on it related to this exact thing with steps to take including how to apply to get renewed access to the device check API for phone auth recaptcha avoidance. I think someone has put it in a related issue here as well. If you find it and paste the link here it would be a service

**Reply by sachin3072004** [reply]
Hi Mike, Thank you for your response. I am not sure if you are referring to the following link. https://github.com/invertase/react-native-firebase/discussions/6974

**Reply by mikehardy** [reply]
Not quite - first - note the title of this discussion is *appcheck* related. You are discussing Auth, so it's not really on target for this discussion.

Play Integrity is a new system to verify a device + app installation is genuine, and is orthogonal to both AppCheck and Auth.
SafetyNet is the old system to verify a device + app installation is genuine, and is orthogonal to both AppCheck and Auth.

AppCheck can use *either* SafetyNet or Play Integrity to get an AppCheck token.

Firebase Auth historically has required SafetyNet to attest an app is genuine in order to avoid the reCAPTCHA. I do not believe it supports Play Integrity yet.

firebase-android-sdk will have to support Play Integrity (if it doesn't? have you asked on firebase-android-sdk repo - might be worthwhile) in order to avoid the reCAPTCHA on android now that SafetyNet is no longer allowed for projects.

Or, there is apparently a form you can submit to google related to that API on the google cloud console and they'll review the request and possibly allow you to continue using SafetyNet until firebase auth supports it.

Unfortunately, I'm not aware of anything actionable we can do here at the react-native-firebase layer. This is entirely handled at the firebase-android-sdk layer underneath us

**Reply by mikehardy** [reply]
firebase-android-sdk BoM 31.4.0 supports play integrity as the default method for phone auth, I'm working on a PR for that now and will release it shortly

**Comment by ledinhthi** [comment]
i’m still get stuck to get play intergrity token from our app on play store
:((. ẽchang always be failed
Vào Th 7, 25 thg 3, 2023 lúc 20:07 Mike Hardy ***@***.***> đã
viết:

**Comment by ledinhthi** [comment]
Great, thanks for the update.

Vào Th 5, 6 thg 4, 2023 lúc 05:06 Mike Hardy ***@***.***> đã
viết:

**Comment by dev007github** [comment]
Hi,
lots of otp auth issues facing in android (ios working fine) after upgrading package to the latest version. please help..

previously used version
    "@react-native-firebase/app": "^17.3.0",
    "@react-native-firebase/app-check": "^17.0.0",
    "@react-native-firebase/auth": "^16.1.1",

current version 
     "@react-native-firebase/app": "^17.4.2",
    "@react-native-firebase/app-check": "^17.4.2",
    "@react-native-firebase/auth": "^17.4.2",

app check code 

**let rnfbProvider = firebase
        .appCheck()
        .newReactNativeFirebaseAppCheckProvider();

      rnfbProvider.configure({
        android: {
          provider: 'playIntegrity',
        },
        apple: {
          provider: 'appAttestWithDeviceCheckFallback',
        },
        web: {
          provider: "reCaptchaV3",
          siteKey: "unknown",
        },
      });

      firebase
        .appCheck()
        .initializeAppCheck({
          provider: rnfbProvider,
          isTokenAutoRefreshEnabled: true,
        });**



**issue:**
in simulator some times getting error like "this app is not authorized to use firebase authentication...." but already registered playintegrity in firebase and added sha1 and sha64... some times auth working with browser recaptcha flow.

in real device auth working with browser recaptcha flow.. but previously otp sms format was currect in release version.. but now otp contains firebase project url in release version..

My previous app version is still working fine for otp auth...so I tried to revert above package versions.. but no luck..same issue...
any solution for this issue?

**Reply by Sameed1998** [reply]
Have you integrated the play integrity api in react native?

**Comment by hanykumar** [comment]
Hi, I face issue implementing Play Integrity App check in bare react native project.

I added this in app.tsx file. 

const appCheckFunction = async () => {
    try {
      let rnfbProvider = AppCheck().newReactNativeFirebaseAppCheckProvider();
      
      rnfbProvider.configure({
        android: {
          provider: 'playIntegrity',
        },
      });
  
      await AppCheck().initializeAppCheck({
        provider: rnfbProvider,
        isTokenAutoRefreshEnabled: true,
      });
  
  
      const appCheckToken = await AppCheck().getToken();
      console.log('AppCheck Token:', appCheckToken);
    } catch (error) {
      console.error('Error initializing or getting AppCheck token:', error);
    }
  };
  
  appCheckFunction();
  
  
added deps in app/build.gradle file. 
    implementation platform('com.google.firebase:firebase-bom:32.7.0')
    implementation("com.google.firebase:firebase-appcheck")
    implementation("com.google.android.play:integrity:1.3.0")
    implementation("com.google.firebase:firebase-appcheck-playintegrity")

google-service.json file is included. 
playIntegration is enabled in firebase as well with sha keys.


But still it gives error 
 ERROR  Error initializing or getting AppCheck token: [Error: [appCheck/token-error] com.google.firebase.FirebaseException: Error returned from API. code: 403 body: App attestation failed.]
 
 
on this line
 const appCheckToken = await AppCheck().getToken();
 console.log('AppCheck Token:', appCheckToken);
 
Any help would be grateful.
Thanks

**Reply by mikehardy** [reply]
Hi there!

```

implementation platform('com.google.firebase:firebase-bom:32.7.0')
implementation("com.google.firebase:firebase-appcheck")
implementation("com.google.android.play:integrity:1.3.0")
implementation("com.google.firebase:firebase-appcheck-playintegrity")
```

Please don't do that. `@react-native-firebase/app` expects to control the `firebase-bom` version, and `@react-native-firebase/app-check` controls all the needed dependencies for app-check.

You should remove all of those dependency lines.


> ERROR Error initializing or getting AppCheck token: [Error: [appCheck/token-error] com.google.firebase.FirebaseException: Error returned from API. code: 403 body: App attestation failed.]

That means app attestation failed. There are many reasons for that. Did you see anything else in the logs? Have you made sure the APIs are enabled on the google side? This looks like it will likely be a project configuration error vs a react-native-firebase error

**Reply by sunilanumolu** [reply]
@hanykumar hey, did u find any solution for this issue? TIA
